File: Java/leetcode75/src/main/java/org/phoenix/leetcode/challenges/Problem57_LetterCombinationsOfAPhoneNumber.java
package org.phoenix.leetcode.challenges;

import java.util.ArrayList;
import java.util.List;

public class Problem57_LetterCombinationsOfAPhoneNumber {
    public List<String> letterCombinations(String digits) {
        List<String> resList = new ArrayList<>();
        if (digits == null || digits.length() == 0) {
            return resList;
        }
        String[] mappings = {"0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

        helper(resList, digits, mappings, 0, "");
        return resList;
    }

    private void helper(List<String> resList, String digits, String[] mappings, int index, String currString) {
        if (index == digits.length()) {
            resList.add(currString);
            return;
        }
        String letters = mappings[digits.charAt(index) - '0'];
        for (int i = 0; i < letters.length(); i++) {
            helper(resList, digits, mappings, index + 1, currString + letters.charAt(i));
        }
    }
}

File: JavaScript/leetcode75/src/main/Problem57_LetterCombinationsOfAPhoneNumber.js
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function (digits) {
  if (digits == null || digits.length === 0) return [];

  const map = {
    2: "abc",
    3: "def",
    4: "ghi",
    5: "jkl",
    6: "mno",
    7: "pqrs",
    8: "tuv",
    9: "wxyz",
  };

  const res = [];
  const helper = (i, s) => {
    if (i === digits.length) {
      res.push(s);
      return;
    }

    for (const c of map[digits[i]]) {
      helper(i + 1, s + c);
    }
  };

  helper(0, "");
  return res;
};

module.exports = letterCombinations;
